### DISCUSSION TITLE
How to smooth lines for antialiasing？

### DISCUSSION THREAD

**Discussion by yexiang92** [pyvista/pyvista#4494] [Q&A]
Hello! One confusion I've always had is how pyvista smoothes lines for anti-aliasing, this is a comparison of plots generated by pyvista and mayavi. In pyvista, I set “plotter.enable_anti_aliasing("msaa")”, but there still seems to be a lot of jaggies. How can I get the effect similar to mayavi? Thanks!
mayavi:
![mayavi](https://s2.loli.net/2023/06/06/NokuGDqeaAWSJ8h.png)
pyvista:
![pyvista](https://s2.loli.net/2023/06/06/cl1gtxXDFYh9bAZ.png)

**Comment by banesullivan** [comment]
Would you be able to share a minimal-ish example to plot with both Mayavi and PyVista for us to compare? Once we have a comparison example, we should be able to see what rendering parameters Mayavi is using and reproduce it

**Comment by akaszynski** [comment]
Consider using `"ssaa"` for the best possible anti-aliasing at the cost of increased render times. See [PyVista - Anti-Aliasing](https://docs.pyvista.org/version/stable/examples/02-plot/anti-aliasing.html).

```py
import pyvista as pv
from pyvista import examples

bunny = examples.download_bunny()
cpos = [(-0.08566, 0.18735, 0.20116), (-0.05332, 0.12168, -0.01215), (-0.00151, 0.95566, -0.29446)]

pl = pv.Plotter()
pl.add_mesh(bunny, show_edges=True, line_width=2)  # lines are thinner in SSAA
pl.enable_anti_aliasing('ssaa')
pl.camera_position = cpos
pl.show()
```

![image](https://github.com/pyvista/pyvista/assets/11981631/6c87cb6b-da1c-41c7-8c52-98e71f270ddc)

**Reply by yexiang92** [reply]
@banesullivan @akaszynski 
Hello! I wrote a short code to demonstrate the difference between pyvista and mayavi. After using `enable_anti_aliasing`, the lines of pyvista are much smoother, but I still have some problems:
(1) The width of the line using the "ssaa" method is very thin;
(2) Mayavi seems to have anti-aliasing by default, and I don't quite understand their internal settings.
(3) pyvista and mayavi render the same hexadecimal color differently.
(4) When dealing with a large number of lines, it seems that mayavi's lines are smoother (default), how can pyvista be improved?

Here is the sample code:

```python
    from mayavi import mlab
    import numpy as np
    import pyvista as pv
    from matplotlib.colors import to_rgb
    
    nodes = np.array([
        [0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0],
        [4.0, 3.0, 0.0],
        [4.0, 0.0, 0.0],
        [0.0, 1.0, 2.0],
        [4.0, 1.0, 2.0],
        [4.0, 3.0, 2.0],
    ])
    edges = np.array(
        [
            [0, 4],
            [1, 4],
            [3, 4],
            [5, 4],
            [6, 4],
            [3, 5],
            [2, 5],
            [5, 6],
            [2, 6],
        ]
    )
    
    ##################################################
    ######################   pyvista
    pv.set_plot_theme('document')
    padding = np.empty(edges.shape[0], int) * 2
    padding[:] = 2
    edges_w_padding = np.vstack((padding, edges.T)).T
    
    mesh = pv.PolyData(nodes, lines=edges_w_padding)
    
    plotter = pv.Plotter()
    plotter.add_mesh(mesh, color="#037ef3", line_width=10,
                     render_lines_as_tubes=True)
    plotter.enable_anti_aliasing("ssaa")
    plotter.show()
    
    ##################################################
    ##############################   mayavi
    mlab.figure(bgcolor=(1, 1, 1))
    src = mlab.pipeline.scalar_scatter(nodes[:, 0], nodes[:, 1], nodes[:, 2], nodes[:, 0]*0)
    src.mlab_source.dataset.lines = edges
    src.update()
    # Finally, display the set of lines
    mlab.pipeline.surface(src, color=to_rgb("#037ef3"), line_width=5, opacity=1.0)
    # lines = mlab.pipeline.stripper(src)
    # lines = mlab.pipeline.tube(src, tube_radius=0.04)
    # mlab.pipeline.surface(lines, color=to_rgb("#037ef3"), line_width=5, opacity=1.0)
    mlab.show()`
 ```

pyvista without enable_anti_aliasing
![pyvista without enable_anti_aliasing](https://s2.loli.net/2023/06/14/vXNgE3z9DMjGmcZ.png)

pyvista with enable_anti_aliasing("msaa")
![pyvista with enable_anti_aliasing("msaa")](https://s2.loli.net/2023/06/14/6r85ZfyiIJl7GKN.png)

pyvista with enable_anti_aliasing("ssaa")
![pyvista with enable_anti_aliasing("ssaa")](https://s2.loli.net/2023/06/14/2wnDsdkfoZIFjEx.png)

mayavi without line as tube
![mayavi without line as tube](https://s2.loli.net/2023/06/14/3ZMWuzasCF5y4Ab.png)

mayavi with line as tube
![mayavi with line as tube](https://s2.loli.net/2023/06/14/lnJFWpzb4U5C6EO.png)

**Reply by banesullivan** [reply]
Just a quick note about the color differences: I think this is due to lighting differences

**Reply by akaszynski** [reply]
I've noticed that SSAA looks the best, but it ends up making my lines look much narrower. I tend to increase the thickness when enabling SSAA, especially when they're initially very narrow.

**Reply by yexiang92** [reply]
@akaszynski @banesullivan 
Thanks a lot for both answers. Since I frequently use beam elements in finite elements, they are all visualized with lines. May I ask what other parameters I may have overlooked to better control the smoothness of the display?
